src/ngraph_encapsulate_op.cc

```diff
@@ -292,6 +292,59 @@ class NGraphEncapsulateOp : public OpKernel {
 
       auto function_size = ng_function->get_graph_size() / 1024;  // kb unit
 
+      // Get the serialized ops and stored the allreduce ops to a vector and
+      vector<shared_ptr<ng::Node>> allreduce_op_list;
+      for (const shared_ptr<ng::Node>& node : ng_function->get_ordered_ops()) {
+        if (node->get_name().compare(0, 9, "AllReduce") == 0) {
+          allreduce_op_list.push_back(node);
+          // Diagnosis to check the input, dependencies, and output for each
+          // Allreduce op
+          if (std::getenv("NGRAPH_DISABLE_LOGGING") == nullptr) {
+#if defined NGRAPH_DISTRIBUTED
+            ngraph::Distributed dist;
+            int Rank_ID;
+            Rank_ID = dist.get_rank();
+            std::cout << "NGTF_Rank: " << Rank_ID << "  " << node->get_name()
+                      << "(";
+            vector<string> inputs;
+            for (const ng::descriptor::Input& input : node->get_inputs()) {
+              inputs.push_back(input.get_tensor().get_name());
+            }
+            std::cout << ng::join(inputs);
```

Comment:
I assume we should use some debug log like `NGRAPH_VLOG` instead of `cout` for this logging messages. The same applies to logging message below.

Reply:
I would like to get all the output in a single line so it is easy to do grep analysis. Tried `NGRAPH_VLOG` and it outputs separated lines. Not sure how to fix that

Reply:
@avijit-nervana any thought?

Reply:
I think if it helps to debug the race conditions then it's ok to dump to stdout. However, please change the following:
1. Package the entire AllReduce processing into a separate function and call it from here if it's a distributed build. 
2. Within the function if you want you may use a different debug flag to enable/disable the logging to stdout

Reply:
@jianyinglang You can create a debug_string and then print it using NGRAPH_VLOG.
stringstream is one way
https://github.com/NervanaSystems/ngraph-tf/blob/master/src/ngraph_encapsulate_op.cc#L263